libraries/AP_Math/SplineCurve.cpp

```diff
@@ -0,0 +1,233 @@
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU General Public License for more details.
+
+ You should have received a copy of the GNU General Public License
+ along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Math/AP_Math.h>
+#include <AP_InternalError/AP_InternalError.h>
+#include "SplineCurve.h"
+
+extern const AP_HAL::HAL &hal;
+
+#define SPEED_MIN           50.0f   // minimum speed in cm/s
+#define ACCEL_MIN           50.0f   // minimum acceleration in cm/s/s
+#define LEASH_LENGTH_MIN    100.0f  // minimum leash length in cm
+#define TANGENTIAL_ACCEL_SCALER 0.5f    // the proportion of the maximum accel that can be used for tangential acceleration (aka in the direction of travel along the track)
+#define LATERAL_ACCEL_SCALER    1.0f    // the proportion of the maximum accel that can be used for lateral acceleration (aka crosstrack acceleration)
+
+// limit the maximum speed along the track to that which will achieve a cornering (aka lateral) acceleration of LATERAL_SPEED_SCALER * acceleration limit
+
+// constructor
+SplineCurve::SplineCurve() :
+    _speed_xy_cms(SPEED_MIN),
+    _speed_up_cms(SPEED_MIN),
+    _speed_down_cms(SPEED_MIN),
+    _accel_xy_cmss(ACCEL_MIN),
+    _accel_z_cmss(ACCEL_MIN)
+{
+}
+
+// set maximum speed, acceleration and leash lengths
+void SplineCurve::set_speed_accel(float speed_xy_cms, float speed_up_cms, float speed_down_cms,
+                                   float accel_xy_cmss, float accel_z_cmss)
+{
+    _speed_xy_cms = MAX(speed_xy_cms, SPEED_MIN);
+    _speed_up_cms = MAX(speed_up_cms, SPEED_MIN);
+    _speed_down_cms = MAX(fabsf(speed_down_cms), SPEED_MIN);
+    _accel_xy_cmss = MAX(accel_xy_cmss, ACCEL_MIN);
+    _accel_z_cmss = MAX(accel_z_cmss, ACCEL_MIN);
+}
+
+// set origin and destination using position vectors (offset from EKF origin in cm
+// origin_vel is vehicle velocity at origin (in cm/s in NEU frame)
+// destination_vel is vehicle velocity at destination (in cm/s in NEU frame)
+void SplineCurve::set_origin_and_destination(const Vector3f &origin, const Vector3f &destination, const Vector3f &origin_vel, const Vector3f &destination_vel)
+{
+    // store origin and destination locations
+    _origin = origin;
+    _destination = destination;
+    _origin_vel = origin_vel;
+    _destination_vel = destination_vel;
+    _reached_destination = false;
+
+    // reset time
+    // Note: _time could include left-over from previous waypoint
+    _time = 0.0f;
+
+    // code below ensures we don't get too much overshoot when the next segment is short
+    const float vel_len = _origin_vel.length() + _destination_vel.length();
+    const float pos_len = (_destination - _origin).length() * 4.0f;
+    if (vel_len > pos_len) {
+        // if total start+stop velocity is more than twice position difference
```

Comment:
Yes, this is because the spline is not fundermentaly a kinematic curve. It is a geometric curve whose parameters are derived from our velocities and positions.